### PR TITLE
perf: optimize SelectMergeProfile

### PR DESCRIPTION
**Pull request by kolesnikovae**
The PR:
 - Enables compression for `SelectMergeProfile` and `SelectMergeStacktrace` handlers.
 - Optimizes reading of deduplicated blocks `SelectMergeProfile` (skips filtering).
 - Enables query split by interval for `SelectMergeProfile`.
 - Minor improvements, such as use of VTProto marshal/unmarshal methods instead of the default ones.
 - Adds ability to truncate pprof profiles (`max_node` parameter added to `SelectMergeProfile`).

---

My initial goal was to optimize `SelectMergeProfile` and use it instead of `SelectMergeStacktraces` in order to get location information (such as line number, file name, etc) in the flamegraph.

However, this does not seem possible due to the vast overhead pprof brings: while a tree / famegraph has compression ratio 10 (gzip), with pprof we can barely expect compression ratio 2. This makes it virtually impossible to switch to pprof without a performance degradation. Moreover, without truncation, we easily hit the default gRPC message size limit (100MB):

image 

Besides, there are certain limitations of the pprof processing flow:
 - In a single querier/query-frontend, we can handle ~50 pprof profiles per second (max_nodes is 8K, ca. 2MB raw uncompressed pprof) per core. With the tree representation (used for building flamegraphs), the rate is appx. 350 profiles per second. Optimisation of this procedure is not trivial and will likely require us to use an alternative format.
 - A single querier frontend merges responses from all the queriers, therefore we can't scale them (querier and query-frontend services) horizontally efficiently.

This makes me thinking that we probably first have to introduce "function selector" that allows to scope a flamegraph to a specific location – basically, request a new flamegraph with _more details_ where the selected node is the root (not to confuse with the existing "Focus block" view). This "drill-down" effect will allow us to set a lower `max_nodes` limit (2-4-8K should be fine) as it mitigates the negative impact of truncation.

Nevertheless, exporting data in pprof will still be limited to the max message size (on the querier / query-frontend side). I don't think this is a problem, because I know no tools capable of handling this big profiles; but we should probable handle this more gracefully.

### CODE REVIEW

**Review by cyriltovena — APPROVED** [review]
LGTM

File: api/ingester/v1/ingester.proto [hunk 1]
```diff
@@ -177,6 +177,9 @@ message MergeProfilesPprofRequest {
   // The client starts the stream with a request containing the profile type and the labels.
   SelectProfilesRequest request = 1;
 
+  // Max nodes in the resulting profile.
+  optional int64 max_nodes = 3;
```
Comment (aleks-p) [inline api/ingester/v1/ingester.proto:181]: a naive question perhaps, should this go below profiles so that they are in order?

**Review by aleks-p — APPROVED** [review]
LGTM as well! I left a few low priority questions / comments.

File: pkg/model/flamegraph.go [hunk 2]
```diff
@@ -246,3 +247,25 @@ func buildStack(dst []string, f *querierv1.FlameGraph, level, idx int) []string
 	}
 	return dst
 }
+
+func TreeFromPprof(profile *profilev1.Profile, sampleIdx int) *Tree {
```
[the remark this answers is not in the document]
Comment (kolesnikovae) [reply pkg/model/flamegraph.go]: Yeah, but this is a test helper itself :) Moved it to the place it is used (pkg/phlaredb/symdb/resolver_test.go)

File: api/ingester/v1/ingester.proto (hunk 1 above)
Comment (kolesnikovae) [reply api/ingester/v1/ingester.proto:180]: I decided to put fields in this order for consistency with other requests incapsulating `SelectProfilesRequest` and to logically separate options specific to `MergeProfilesPprofRequest` and the profile batch which is the only member we actually stream

File: api/querier/v1/querier.proto [hunk 3]
```diff
@@ -114,6 +114,7 @@ message SelectMergeProfileRequest {
   string label_selector = 2;
   int64 start = 3; // milliseconds since epoch
   int64 end = 4; // milliseconds since epoch
+  optional int64 max_nodes = 6; // Limit the nodes returned to only show the node with the max_node's biggest total
```
[the remark this answers is not in the document]
Comment (kolesnikovae) [reply api/querier/v1/querier.proto]: There is a [great resource](https://protobuf.dev/programming-guides/proto3/#assigning) explaining how fields should be defined and numbered, and [a lower-level overview](https://protobuf.dev/programming-guides/encoding/#order) of the encoding approaches

[the remark this answers is not in the document]
Comment (aleks-p) [reply api/querier/v1/querier.proto]: thank you!

**Pull request merged**